dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,413 @@
+    INSTRUMENTER.finalize(metadata).map_err(|e| eyre!(e))?;
+    Ok(())
+}
+
+/// Delete all files that we want to be regenerated
+/// so that we always have fully-up-to-date, non-incremental metadata.
+fn delete_metadata_and_dependencies(
+    info: &InstrumentInfo,
+    cargo_metadata: &Metadata,
+) -> eyre::Result<()> {
+    if let Err(e) = fs_err::remove_file(&info.metadata) {
+        if e.kind() != ErrorKind::NotFound {
+            return Err(e.into());
+        }
+    }
+
+    // TODO(kkysen) Figure out a way to know which profile is being used
+    // Until then, just search them all and delete all of them.
+
+    // TODO(kkysen) We probably have to delete binaries that have different names from the crates.
+
+    // Delete all executables in `target/${profile}/deps/` starting with the crate name + `-`.
+    let target_read_dir = match cargo_metadata.target_directory.read_dir() {
+        Ok(read_dir) => Ok(read_dir),
+        Err(e) => if e.kind() == ErrorKind::NotFound {
+            return Ok(());
+        } else {
+            Err(e)
+        }
+    }?;
+    for profile_dir in target_read_dir {
+        let profile_dir = profile_dir?;
+        if !profile_dir.file_type()?.is_dir() {
+            continue;
+        }
+        for artifact in profile_dir.path().join("deps").read_dir()? {
+            let artifact = artifact?;
+            if !artifact.file_type()?.is_file() {
+                continue;
+            }
+            let file_name = artifact.file_name();
+            let file_name = file_name.to_str();
+            for crate_target in &info.crate_targets {
+                let prefix = format!("{}-", crate_target.crate_name.as_str());
+                // [`Path::starts_with`] only checks whole components at once,
+                // and `OsStr::starts_with` doesn't exist yet.
+                if file_name
+                    .map(|name| name.starts_with(&prefix))
+                    .unwrap_or_default()
+                {
+                    let artifact = artifact.path();
+                    if artifact.is_executable() {
+                        fs_err::remove_file(&artifact)?;
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+    env_logger::init();
+
+    let rustc_wrapper_var = "RUSTC_WRAPPER";
+    let instrument_info_var = "C2RUST_INSTRUMENT_INFO";
+
+    let own_exe = env::current_exe()?;
+
+    let wrapping_rustc = env::var_os(rustc_wrapper_var).as_deref() == Some(own_exe.as_os_str());
+    if wrapping_rustc {
+        let sysroot = get_sysroot()?;
+        let at_args = env::args().skip(1).collect::<Vec<_>>();
+        let crate_target = CrateTarget::from_rustc_args(&at_args)?;
+        let info = env::var(instrument_info_var)?;
+        let info = serde_json::from_str::<InstrumentInfo>(&info)?;
+        let should_instrument = info.crate_targets.contains(&crate_target);
```

Comment:
Cargo sets `CARGO_PRIMARY_PACKAGE` when compiling the primary package (i.e. not dependencies). Could we use this instead of passing info and parsing the command line?

Reply:
I don't remember seeing this environment variable (I dumped and looked through all of the ones passed to the `rustc` wrapper), but I can check.  If it is, I think we can use it instead of the current way I have of checking if we should instrument.

Reply:
`$CARGO_PRIMARY_PACKAGE` works, which allows things to be simplified a ton.  Thanks!

Fixed in:
* https://github.com/immunant/c2rust/pull/554/commits/a1fdf172fd7abaaf387c69a949748d68f2c22cba
* https://github.com/immunant/c2rust/pull/554/commits/d28775ae0f5a1404ee8f24e2d570e3a37d25fc6c
* https://github.com/immunant/c2rust/pull/554/commits/8c5568d3e1d2c0a2136306a0a81df4dd11663e89